File: mobile/H3MShop/Model/Bill.swift
//
//  Bill.swift
//  H3MShop
//
//  Created by Lê Văn Huy on 07/11/2022.
//

import Foundation

class BillRequest: Codable,Identifiable{
    var product: [ProductBill]
    var payment_method:String
    var delivery: String
    var weight: Double
    var productPrice: Int
    var shipPrice: Int
    var id_info : String
    
    init(product: [ProductBill] = [], payment_method: String, delivery: String, weight: Double, productPrice: Int, shipPrice: Int, id_info: String) {
        self.product = product
        self.payment_method = payment_method
        self.delivery = delivery
        self.weight = weight
        self.productPrice = productPrice
        self.shipPrice = shipPrice
        self.id_info = id_info
    
    }
}

class ProductBill: Codable,Identifiable{
    var id_product: String
    var size: String
    var color: String
    var number: Int
    var price: Int
    
    init(id_product: String , size: String, color: String, number: Int, price: Int) {
        self.id_product = id_product
        self.size = size
        self.color = color
        self.number = number
        self.price = price
    }
}







class SizeBill: Codable,Identifiable {
    var _id: String
    var name: String
    
    init(_id: String = "" ,name: String = "" ){
        self._id = _id
        self.name = name
    }
}

class ColorBill: Codable,Identifiable {
    var _id: String
    var name: String
    
    init(_id: String = "" ,name: String = "" ){
        self._id = _id
        self.name = name
    }
}


//MARK: id_product
class ProductHistory: Codable {
    var _id: String
    var name: String
    var urlImage:[String]
}


//MARK: product
class ProductBillResponse: Codable,Identifiable{
    var _id: String
    var id_product: ProductHistory
    var size: SizeBill
    var color: ColorBill
    var number: Int
    var price: Int
}

class InfoAddressBill: Codable,Identifiable{
    
    var _id: String
    var address: AddressDetail
    var name: String
    var phone: String
    
    init(_id: String = "",address: AddressDetail = AddressDetail(), name: String = "" , phone: String = "") {
        self._id = _id
        self.address = address
        self.name = name
        self.phone = phone
    }
}





class HistoryBill:Codable,Identifiable {
    var _id: String
    var product: [ProductBillResponse]
    var payment_method: PaymentMethodBill?
    var delivery: DeliveryBill
    var totalPrice: Int
    var productPrice: Int
    var shipPrice: Int
    var createAt: String
    var id_info: InfoAddressBill
    
    init(_id: String = "" , product: [ProductBillResponse] = [], payment_method: PaymentMethodBill = PaymentMethodBill(), totalPrice: Int = 0, productPrice: Int = 0, shipPrice: Int = 0, createAt: String = "", id_info: InfoAddressBill = InfoAddressBill(),delivery: DeliveryBill = DeliveryBill()) {
        self._id = _id
        self.product = product
        self.payment_method = payment_method
        self.totalPrice = totalPrice
        self.productPrice = productPrice
        self.shipPrice = shipPrice
        self.createAt = createAt
        self.id_info = id_info
        self.delivery = delivery
    }
}

class PaymentMethodBill: Codable,Identifiable {

    var _id: String
    var name: String
    
    init(_id: String = "" , name: String = "" ) {
        self._id = _id
        self.name = name
    }
}

class DeliveryBill: Codable{
    var _id: String
    var name: String
    
    init(_id: String = "" , name: String = "" ) {
        self._id = _id
        self.name = name
    }
}


class History: Codable,Identifiable {
    var id_status: Status
    var date: String
    var _id: String
    
}

class Status: Codable {
    var _id: String
    var name: String
    
    init(_id: String = "", name: String = "") {
        self._id = _id
        self.name = name
    }
    
}


class Bill: Codable,Identifiable {
    var _id: String
    var id_bill: HistoryBill
    var history : [History]
    
    init(_id: String = "" , id_bill: HistoryBill, history: [History] = []) {
        self._id = _id
        self.id_bill = id_bill
        self.history = history
    }
}





class BillResponse: Codable {
    var message: String
    var bill: [Bill]
}

File: api/Route/Bill.js
const express = require("express");
const auth = require("../middleware/auth");
const { populate } = require("../Model/Account");
const { Bill } = require("../Model/Bill");
const { Cart } = require("../Model/Cart");
const { Order_history } = require("../Model/Order_history");
const { Stock } = require("../Model/Stock");
const router = express.Router();
var schedule = require("node-schedule");
const { StockProduct } = require("../Model/StockProduct");
// var j = schedule.scheduleJob(myToday, function(){
//   console.log('hello world.');
// });
async function UpdateStockProduct(product) {
  const number = product.number * -1;
  let stockProduct = await StockProduct.findOneAndUpdate(
    {
      id_product: product.id_product,
      size: product.size,
      color: product.color,
    },
    { $inc: { number: number } }
  );
}

router.post("/insert", auth, async function (req, res) {
  try {
    let totalPrice = req.body.shipPrice + req.body.productPrice;
    let bill = new Bill({
      delivery: req.body.delivery,
      id_account: req.user.id,
      product: req.body.product,
      shipPrice: req.body.shipPrice,
      productPrice: req.body.productPrice,
      totalPrice: totalPrice,
      createAt: Date.now(),
      id_info: req.body.id_info,
      payment_method: req.body.payment_method,
    });
    await bill.save();
    bill.product.forEach(async function (product) {
      await UpdateStockProduct(product);
    });
    await Cart.findOneAndUpdate(
      { id_account: req.user.id },
      { $set: { product: [] } }
    );
    let orderhistory = new Order_history({
      id_account: req.user.id,
      id_bill: bill._id,
      delivery: bill.delivery,
      history: [{ id_status: "63691e673f2070927236ba3f", date: Date.now() }],
    });
    orderhistory = await orderhistory.save();
    res.status(200).send({ message: "success", status: true });
    const d = new Date();
    d.setHours(d.getMinutes + 1);
    schedule.scheduleJob(d, function () {
      console.log("hello world.");
      // });
    });
  } catch (ex) {
    res.status(400).send({ message: "error", status: false });
    console.log(ex);
  }
});

router.get("/all", auth, async function (req, res) {
  console.log("3");
  try {
    let orderhistory = await Order_history.find({
      id_account: req.user.id,
      isCancel: { status: false },
    })
      .select(["id_bill", "history"])
      .populate({
        path: "id_bill",
        select: [
          "id_info",
          "product",
          "totalPrice",
          "createAt",
          "productPrice",
          "shipPrice",
          "payment_method",
          "delivery",
        ],
        populate: [
          { path: "product.id_product", select: ["name", "urlImage"] },
          { path: "product.size", select: "name" },
          { path: "product.color", select: "name" },
          { path: "payment_method", select: "name" },
          { path: "delivery", select: "name" },
          {
            path: "id_info",
            select: ["name", "phone", "address"],
            populate: [
              { path: "address.id_province", select: "name" },
              { path: "address.id_district", select: "name" },
              { path: "address.id_commune", select: "name" },
            ],
          },
        ],
      })
      .populate("history.id_status", ["name"]);
    res.status(200).send({ message: "success", bills: orderhistory });
  } catch (ex) {
    res.status(400).send({ message: "error", status: false });
    console.log(ex);
  }
});
router.get("/type/:type", auth, async function (req, res) {
  try {
    let filter = {
      id_account: req.user.id,
      isCancel: { status: false },
      history: { $size: req.params.type },
    };
    let select = ["id_bill", "history"];
    if (req.params.type == 5) {
      filter = { id_account: req.user.id, isCancel: { status: true } };
      select = ["id_bill", "history", "isCancel"];
    }
    let orderhistory = await Order_history.find(filter)
      .select(select)
      .populate({
        path: "id_bill",
        select: [
          "id_info",
          "product",
          "totalPrice",
          "createAt",
          "productPrice",
          "shipPrice",
          "payment_method",
        ],
        populate: [
          { path: "product.id_product", select: ["name", "urlImage"] },
          { path: "product.size", select: "name" },
          { path: "product.color", select: "name" },
          { path: "payment_method", select: "name" },
          { path: "delivery", select: "name" },
          {
            path: "id_info",
            select: ["name", "phone", "address"],
            populate: [
              { path: "address.id_province", select: "name" },
              { path: "address.id_district", select: "name" },
              { path: "address.id_commune", select: "name" },
            ],
          },
        ],
      })
      .populate("history.id_status", ["name"]);
    res.status(200).send({ message: "success", bill: orderhistory });
  } catch (ex) {
    res.status(400).send({ message: "error", status: false });
  }
});
router.get("/detail/:id", async function (req, res) {
  try {
    let orderhistory = await Order_history.findById(req.params.id)
      .select(["id_bill", "history"])
      .populate({
        path: "id_bill",
        select: [
          "id_info",
          "product",
          "totalPrice",
          "createAt",
          "productPrice",
          "shipPrice",
          "payment_method",
        ],
        populate: [
          { path: "product.id_product", select: ["name", "urlImage"] },
          { path: "product.size", select: "name" },
          { path: "product.color", select: "name" },
          { path: "payment_method", select: "name" },
          { path: "delivery", select: "name" },
          {
            path: "id_info",
            select: ["name", "phone", "address"],
            populate: [
              { path: "address.id_province", select: "name" },
              { path: "address.id_district", select: "name" },
              { path: "address.id_commune", select: "name" },
            ],
          },
        ],
      })
      .populate("history.id_status", ["name"]);
    res.status(200).send({ message: "success", bill: orderhistory });
  } catch (ex) {
    res.status(400).send({ message: "error", status: false });
  }
});
module.exports = router;
